src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -182,6 +200,26 @@ private static void EmitRegexDerivedTypeRunnerFactory(IndentedTextWriter writer,
                 }
                 writer.WriteLine();
             }
+            if (rm.MatchTimeout is null)
+            {
+                // We need to emit timeout checks for everything other than the developer explicitly setting Timeout.Infinite.
+                // In the common case where a timeout isn't specified, we need to at run-time check whether a process-wide
+                // default timeout has been specified, so we emit a static readonly TimeSpan to store the default value
+                // and a static readonly bool to store whether that value is non-infinite (the latter enables the JIT
+                // to remove all timeout checks as part of tiering if the default is infinite).
+                const string DefaultTimeoutHelpers = nameof(DefaultTimeoutHelpers);
+                if (!requiredHelpers.ContainsKey(DefaultTimeoutHelpers))
+                {
+                    requiredHelpers.Add(DefaultTimeoutHelpers, new string[]
+                    {
+                        $"/// <summary>Default timeout value set in <see cref=\"AppContext\"/>, or <see cref=\"Regex.InfiniteMatchTimeout\"/> if none was set.</summary>",
+                        $"internal static readonly TimeSpan {DefaultTimeoutFieldName} = AppContext.GetData(\"REGEX_DEFAULT_MATCH_TIMEOUT\") is TimeSpan timeout ? timeout : Regex.InfiniteMatchTimeout;",
```

Comment:
why is this calling GetData directly rather than calling  InitDefaultMatchTimeout, which also checks that it's valid? Then REGEX_DEFAULT_MATCH_TIMEOUT is only in one place.

Reply:
Because InitDefaultMatchTimeout is private and this code is in the user's assembly.

Reply:
ah yes, it will fail there if the object is not a TimeSpan, etc.